### DISCUSSION TITLE
Temporary invoice numbers

### DISCUSSION THREAD

**Discussion by sebastiancretu** [crater-invoice-inc/crater#561] [Ideas]
To avoid numbering gaps we need to provide a temporary number as invoice number until this invoice is sent, then it takes the final invoice number with the right sequence. Users should have a setting that allows for the creation of temporary invoices in order to maintain the right consecutive number on all sent invoices. These invoices are also included in sales reports and dashboards.

I think we can use this workflow :

- When creating a new invoice:
  - set a temporary sequence number using the pattern **T-1*** (where 1 is the invoice entry id)
  - set the sequence number to null or 0
  - allow user to edit the invoice number
- When invoice is sent: 
  - replace the temporary invoice number with the right invoice number containing the new sequence if it wasn't set manually by the user on creation.
  - replace the invoice date with the date when the invoice is actually sent
  - replace the due date to count from the date the invoice was sent
- On settings:
  - have a "Temporary Invoices" option. When active we use the above functionality. When inactive the invoice number automatically from the beginning.

When cloning or converting an estimate to invoice, the above functionality will be kept.

For frontend, this types of invoice can be marked with a label.

**Comment by LU-386** [comment]
Great, **thanks for your support** for the idea so far!

I think using the send-date as base for invoice- and due-date is a great idea and i would use it myself, but there should be an option so you can disable it.

I'm not sure if this feature should be the same for estimates and payments, but i think it could be helpful too (unsure about the overhead to implement this additionally, maybe with options for enable/disabling this for each of those.

Otherwise your workflow looks pretty good for me and I hope to see this in one of the next versions.

**Comment by sebastiancretu** [comment]
> I think using the send-date as base for invoice- and due-date is a great idea and i would use it myself, but there should be an option so you can disable it.

Think about this scenario. We create invoice A on date 17.09.2021 and save it. After 2 days we create an invoice B with date 19.09.2021, we save it and send it. If we will send the Invoice A now, it will have a newer number but an older date. I don't think that this should happen.

**Comment by LU-386** [comment]
This is absolutely true, i think the date should be disabled (or at least not editable) if this function is enabled (just to prevent confusion).

Meanwhile it looks like an additional workflow besides the default crater workflow for me, are you able to integrate that in this kinda short circut? Would'nt it better to break it up to "Advanced invoice states" ... but i see you already did that

**Comment by xavsio4** [comment]
Can that be related to "Pro Forma" invoices ?
